Add pr_content_check to features

burtonr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This commit adds a new feature for pr_content_check
This feature will verify that the incoming PR has
something in the description (aka body)

If there is no description, the invalid label will be applied
and a comment will be added to the PR with a link to the
contributing guide.

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md))
Resolves #103 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Unit tests added and passing

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

alexellis commented on USER_GUIDE.md:
```diff
@@ -60,6 +60,10 @@ redirect: https://raw.githubusercontent.com/openfaas/faas/master/.DEREK.yml
 
+### Feature: `no_empty_pr`
+
+If `no_empty_pr` is specified in the feature list then Derek will inform you that a PR needs a description. He also adds the `invalid` label.
```
I'm not sure if `validate_pr_description` / `validate_pr_text` would be more descriptive?

burtonr replied:
Not a problem, the feature name wasn't defined in the issue, so I did my best to succinctly describe the use as the feature name. I was trying to keep the feature name small to reduce the potential for typing errors

rgee0 replied:
`pr_content_check`

burtonr replied:
updated. Concise and clear. Thanks @rgee0!

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -101,6 +86,57 @@ That's something we need before your Pull Request can be merged. Please see our
+	token := os.Getenv("personal_access_token")
+	if len(token) == 0 {
+
+		newToken, tokenErr := auth.MakeAccessTokenForInstallation(
```
`newToken` might be more accurate as `installationToken`

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -101,6 +86,57 @@ That's something we need before your Pull Request can be merged. Please see our
+			installationID,
+			config.PrivateKey)
+
+		if tokenErr != nil {
```
probably better to bubble up an error and handle properly than blow up. I know this may have been refactored out, but that now means it's getting eyes on it. My hope is to remove all `log.Fatal*` calls and `panic` so we can use Derek as an SDK in the future.

burtonr replied:
I wasn't quite sure how the error should be handled. For now, just writing the error to stdout and returning before the token is used to prevent a panic

alexellis replied:
How about returning an error from the function?

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -101,6 +86,57 @@ That's something we need before your Pull Request can be merged. Please see our
+
+	if req.Action == "opened" {
+		if missingDescription(req.PullRequest) {
+			fmt.Println("Applying invalid label")
```
What if we extracted a const for the invalid label so we could change it in the future potentially?

For instance:

```
const noDescriptionLabel = "invalid"
```

burtonr replied:
good idea! updated to a const at the top of the file

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -101,6 +86,57 @@ That's something we need before your Pull Request can be merged. Please see our
+	token := getAccessToken(config, req.Installation.ID)
+	client := factory.MakeClient(ctx, token, config)
+
+	if req.Action == "opened" {
```
is there a `const` defined anywhere for `opened` we could use? I know Richard added a lot of consts earlier and it seemed like a reasonable pattern to follow

burtonr replied:
Good idea! "magic strings" are no fun!

burtonr replied:
const added and updated the places where the string was referenced

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -143,3 +179,7 @@ func hasUnsigned(req types.PullRequestOuter, client *github.Client) (bool, error
 }
+
+func missingDescription(pr types.PullRequest) bool {
+	return len(pr.Body) < 1
```
Not sure this could ever be negative 1 `-1` or less than `0` so how about looking at `== 0`.

One other thought, what if someone pasted this:

```
 
```

(that's a single space)

This may be of use for extending the tests:

```
strings.Trim*
```

burtonr replied:
The `< 1` is my personal standard for testing string length since some languages return a -1 when the value cannot be determined. I know that's not the case in Go, just a habit of mine.

For the blank space, I did test this with Github, and found that they trim pre and post spaces before the request is sent, so I didn't add the additional step. It wouldn't hurt, so I'll add that in just in case.

burtonr replied:
Not sure I follow.
Are you saying this check should validate that there is a word in the description? Rather than just testing the length.

rgee0 replied:
I was thinking that by using a regex we could omit whitespace.  On reflection there will be all manner of non-alphanumerics as its markdown, so let's consign that suggestion to the heap.

ivanayov replied:
What can be done is to check the commit body too, as it's usually automatically added as description.
Three problems with that are
* Still does not guarantee non-empty description
* Moves the scope of the check
* Complicates for multiple commits

But it's still an option.

alexellis replied:
The scope of this issue is only for non-empty bodies. Please see the issue description.

ivanayov replied:
Ok, I just thought for the case, when the body is not empty and still there is no description. But this can be out of scope for this PR.

alexellis replied:
I agree, it's encapsulated in another issue - this one has a much more defined scope.

burtonr replied:
I've updated this piece with the input provided. Added a trim, and check for 0.

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -101,6 +86,57 @@ That's something we need before your Pull Request can be merged. Please see our
+				log.Fatalf("%s limit: %d, remaining: %d", assignLabelErr, res.Limit, res.Remaining)
+			}
+
+			body := `Thank you for your contribution. I've just checked and your Pull Request doesn't appear to have any description.
```
Seems consistent with the previous messages we used.

@rgee0 do you have any input on this?

burtonr replied:
I don't quite have that English "personality", so I'd appreciate input into proper phrasing :smile: 
I know Derek's personality is one of his strong points!

rgee0 replied:
The two comments are very similar.  Is there a way we can extract the main part and just update the 'problem detail' when using it?

`Thank you for your contribution. I've just checked and <problem detail>.
That's something we need before your Pull Request can be merged. Please see our [contributing guide](` + contributingURL + `).`

burtonr replied:
If it's ok, we could do this in a follow up PR. I think there's more to think about than this one instance.

I may have a solution you can see here: https://play.golang.org/p/FByvDuUR4g2
Not sure where the `struct` would be defined though. I was thinking in a new file within the `handlers` package so that it could be used in other handlers.

alexellis commented on main.go:
```diff
@@ -20,9 +20,10 @@ import (
-	deleted  = "deleted"
+	dcoCheck  = "dco_check"
+	comments  = "comments"
+	deleted   = "deleted"
```
Maybe the `opened` const could go here if it's not present already?

burtonr replied:
added the const in the `pullRequestHandler.go` file as that's the only place it's used. Similar to these consts which are only used in the `main.go` file.

Happy to move it if you'd prefer all the const definitions are in one place.

alexellis commented on handler/pullRequestHandler.go:
```diff
@@ -143,3 +196,7 @@ func hasUnsigned(req types.PullRequestOuter, client *github.Client) (bool, error
 	return strings.Contains(msg, "Signed-off-by:")
 }
+
+func missingDescription(pr types.PullRequest) bool {
```
Can you call this `hasDescription` please?

burtonr replied:
updated. Changing the name redefines this function, so I've updated it to return `> 0` rather than the previous implementation of `== 0`

I've updated the calling function and the tests to reflect this new change

alexellis commented:
This is not strictly a `pr_content_check` (the PR being a series of commits - we don't check them at all, and this doesn't speak to how we check them) - I'd prefer to see this more descriptive such as: `pr_description_required` or `mandatory_pr_description`.

Alex

ivanayov commented:
I would vote for `pr_description_required` as it's best descriptive.

burtonr commented:
@alexellis This is ready now. 

I've updated the feature name to the even better, more descriptive: `pr_description_required`
I've addressed your comments

I've also got the version of Derek deployed and installed to test end to end on one of my repos.

PR Missing a description:
![screenshot from 2018-11-22 00-48-50](https://user-images.githubusercontent.com/12687192/48886253-e41f7180-edf0-11e8-866d-35a348c1ec2b.png)

PR with description:
![screenshot from 2018-11-22 00-51-54](https://user-images.githubusercontent.com/12687192/48886263-ef729d00-edf0-11e8-8f45-3a46b2c74046.png)

and finally, a PR with unsigned commits, and no description:
![screenshot from 2018-11-22 00-54-26](https://user-images.githubusercontent.com/12687192/48886337-3d87a080-edf1-11e8-8ff1-862ce292bcd6.png)

martindekov commented:
Before this getting merged I would suggest to add `pr_description_required` in the `.DEREK.yml` file configuration of probably commented out so we know from where to manage this label

martindekov commented:
Yeah ^ it was just suggestion
